cmd/manager/cleanup.go

```diff
@@ -0,0 +1,178 @@
+
+	c.logger.Printf("Updating existing subscriptions.")
+	datastoreCscNames, err := updateSubsciptions(c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting datastore CatalogSourceConfigs.")
+	err = deleteDatastoreCscs(datastoreCscNames, c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting old CatalogSourceConfigs created duing operator installations.")
+	err = deleteOldCscs(c.client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func updateSubsciptions(kubeClient client.Client) ([]string, error) {
+	var cscNames []string
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("csc-owner-name"))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing subscriptions that have the label "csc-owner-name"
+	err := kubeClient.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		return nil, err
+	}
+	for _, instance := range subscriptions.Items {
+		cscNames = append(cscNames, instance.GetLabels()["csc-owner-name"])
+		// try to guess the datastore catalogsource from the subscription
+		datastoreCs, err := findCatalogSource(&instance, kubeClient)
+		if k8s_errors.IsNotFound(err) {
+			// guess the catalogsource from the opsrc that has the package
+			datastoreCs, err = findCsFromOpsrc(&instance, kubeClient)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// point the subscription to the datastore CatalogSource
+		instance.SetCatalogSource(datastoreCs.GetName())
+		instance.SetCatalogSourceNamespace("openshift-marketplace")
+		kubeClient.Update(context.TODO(), &instance)
+	}
+	return cscNames, nil
+}
+
+func deleteDatastoreCscs(cscNames []string, kubeClient client.Client) error {
+	for _, cscName := range cscNames {
+		namespacedName := types.NamespacedName{Name: cscName, Namespace: "openshift-marketplace"}
+		csc := &marketplace.CatalogSourceConfig{}
+		err := kubeClient.Get(context.TODO(), namespacedName, csc)
+		if err == nil {
+			err = kubeClient.Delete(context.TODO(), csc)
+			if err != nil {
+				return err
+			}
+		} else if err != nil && !k8s_errors.IsNotFound(err) {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func deleteOldCscs(kubeClient client.Client) error {
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("opsrc-datastore: \"true\""))
+
+	cscs := &marketplace.CatalogSourceConfigList{}
+	// Get the list of existing cscs that have the lable "opsrc-datastore: "true" "
+	err := kubeClient.List(context.TODO(), options, cscs)
+	if err != nil {
+		return err
+	}
+	for _, csc := range cscs.Items {
+		kubeClient.Delete(context.TODO(), &csc)
+	}
+	return nil
+}
+
+func findCatalogSource(subscription *olm.Subscription, client client.Client) (*olm.CatalogSource, error) {
+	associatedCscName := subscription.GetLabels()["csc-owner-name"]
+	possibleCsName, err := extractCsName(associatedCscName)
+	if err != nil {
+		return nil, err
+	}
+	// try and fetch the CatalogSource
+	datastoreCs := &olm.CatalogSource{}
```

Comment:
Why is this a datastore `CatalogSource`?

Reply:
@aravindhp this is the `CatalogSource` that a subscription will have information about after it is updated. Existing Subscriptions have information about the CSCs installed during operator installation. For example, if a CSC name `installed-community-openshift-operators` was created during installation of an operator, the Subscription had the following information
```
spec:
     source: installed-community-openshift-operators
     sourceNamespace: openshift-operators
```

This has to be changed to contain information about the datastore CatalogSource instead. 

```
spec:
     source: community-operators
     sourceNamespace: openshift-marketplace
```
Given that, shouldn't it be `datastoreCs`?

Reply:
No. Datastore `CatalogSourceConfigs` and `CatalogSources` were created by `OperatorSources` before #198.

Reply:
@aravindhp what should the updated Subscription contain information about in it's `spec.Source` and `spec.SourceNamesapce` fields?

Reply:
> arent datastore CatalogSources being created directly now by OperatorSources?

Correct

> Shouldn't the updated Subscription contain information about the datastore CatalogSources in it's spec.Source and spec.SourceNamesapce fields?

Yes when you update the `Subscription`, it should reference the new datastore `CatalogSource`. The problem here was with some of the function naming that was confusing me. Given you have already pushed changes, I am unable to clarify further. In future, please wait for such conversations to get resolved before you push changes.